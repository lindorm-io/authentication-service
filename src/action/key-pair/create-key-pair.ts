import Joi from "@hapi/joi";
import { IKoaAuthContext } from "../../typing";
import { KeyType } from "@lindorm-io/key-pair";
import { assertAccountAdmin, generateKeyPair } from "../../support";

export interface ICreateKeyPairOptions {
  type: KeyType;
}

export interface ICreateKeyPairData {
  algorithm: string;
  keyPairId: string;
  type: string;
}

const schema = Joi.object({
  type: Joi.string().valid(KeyType.EC, KeyType.RSA).required(),
});

export const createKeyPair = (ctx: IKoaAuthContext) => async (
  options: ICreateKeyPairOptions,
): Promise<ICreateKeyPairData> => {
  await schema.validateAsync(options);

  const { logger, repository } = ctx;
  const { type } = options;

  await assertAccountAdmin(ctx)();

  const keyPair = await generateKeyPair(type);

  await repository.keyPair.create(keyPair);

  logger.debug("key pair created", {
    algorithm: keyPair.algorithm,
    id: keyPair.id,
    type: keyPair.type,
  });

  return {
    algorithm: keyPair.algorithm,
    keyPairId: keyPair.id,
    type: keyPair.type,
  };
};
