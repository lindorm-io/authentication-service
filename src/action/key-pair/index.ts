export * from "./create-key-pair";
export * from "./expire-key-pair";
