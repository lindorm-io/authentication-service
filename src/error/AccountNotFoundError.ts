import { APIError, HttpStatus } from "@lindorm-io/core";

export class AccountNotFoundError extends APIError {
  constructor(accountId?: string) {
    super("Account not found", {
      debug: { accountId },
      statusCode: HttpStatus.ClientError.BAD_REQUEST,
    });
  }
}
