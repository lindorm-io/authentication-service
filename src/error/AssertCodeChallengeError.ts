import { APIError, HttpStatus } from "@lindorm-io/core";

export class AssertCodeChallengeError extends APIError {
  constructor(challenge: string, verifier: string) {
    super("Invalid Code Challenge", {
      debug: { challenge, verifier },
      statusCode: HttpStatus.ClientError.FORBIDDEN,
    });
  }
}
