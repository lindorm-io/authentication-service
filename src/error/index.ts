export * from "./AccountNotFoundError";
export * from "./AssertClientSecretError";
export * from "./AssertCodeChallengeError";
export * from "./AssertDeviceChallengeError";
export * from "./AssertOTPError";
export * from "./DeviceNotFoundError";
export * from "./ExpiredSessionError";
export * from "./InvalidAuthorizationError";
export * from "./InvalidAuthorizationTokenError";
export * from "./InvalidClientError";
export * from "./InvalidDeviceError";
export * from "./InvalidExpiryString";
export * from "./InvalidGrantTypeError";
export * from "./InvalidPermissionError";
export * from "./InvalidRefreshTokenError";
export * from "./InvalidScopeError";
export * from "./InvalidSubjectError";
export * from "./LockedAccountError";
export * from "./RejectedClientError";
export * from "./RequestLimitBackOffError";
