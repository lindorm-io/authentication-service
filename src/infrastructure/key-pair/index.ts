export * from "./KeyPairCache";
export * from "./KeyPairRepository";
