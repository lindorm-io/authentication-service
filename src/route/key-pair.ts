import { BEARER_TOKEN_MW_OPTIONS } from "../config";
import { HttpStatus } from "@lindorm-io/core";
import { IKoaAuthContext } from "../typing";
import { Router } from "@lindorm-io/koa";
import { accountMiddleware } from "../middleware";
import { bearerTokenMiddleware } from "@lindorm-io/koa-jwt";
import { createKeyPair, expireKeyPair } from "../action";
import { clientMiddleware } from "@lindorm-io/koa-client";

export const router = new Router();

router.use(clientMiddleware());
router.use(bearerTokenMiddleware(BEARER_TOKEN_MW_OPTIONS));
router.use(accountMiddleware);

router.post(
  "/",
  async (ctx: IKoaAuthContext): Promise<void> => {
    const { type } = ctx.request.body;

    ctx.body = await createKeyPair(ctx)({ type });
    ctx.status = HttpStatus.Success.CREATED;
  },
);

router.patch(
  "/:id/expire",
  async (ctx: IKoaAuthContext): Promise<void> => {
    const { expires } = ctx.request.body;

    await expireKeyPair(ctx)({
      keyPairId: ctx.params.id,
      expires,
    });

    ctx.body = {};
    ctx.status = HttpStatus.Success.NO_CONTENT;
  },
);
