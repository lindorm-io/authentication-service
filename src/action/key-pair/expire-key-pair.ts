import Joi from "@hapi/joi";
import { IAuthContext } from "../../typing";
import { InvalidExpiryString } from "../../error";
import { add } from "date-fns";
import { assertAccountAdmin } from "../../support";
import { stringToDurationObject, stringToMilliseconds } from "@lindorm-io/core";

export interface IExpireKeyPairOptions {
  keyPairId: string;
  expires: string;
}

const schema = Joi.object({
  keyPairId: Joi.string().guid().required(),
  expires: Joi.string().required(),
});

export const expireKeyPair = (ctx: IAuthContext) => async (options: IExpireKeyPairOptions): Promise<void> => {
  await schema.validateAsync(options);

  const { logger, repository } = ctx;
  const { expires, keyPairId } = options;

  await assertAccountAdmin(ctx)();

  if (stringToMilliseconds(expires) === 0) {
    throw new InvalidExpiryString(expires);
  }

  const keyPair = await repository.keyPair.find({ id: keyPairId });

  keyPair.expires = add(new Date(), stringToDurationObject(expires));

  await repository.keyPair.update(keyPair);

  logger.debug("key pair updated", {
    id: keyPair.id,
    expires: keyPair.expires,
  });
};
