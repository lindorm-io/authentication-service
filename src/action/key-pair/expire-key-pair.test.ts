import MockDate from "mockdate";
import { InvalidExpiryString } from "../../error";
import { KeyPair } from "@lindorm-io/key-pair";
import { expireKeyPair } from "./expire-key-pair";
import { getTestRepository, getTestKeyPairEC, inMemoryStore, resetStore, logger } from "../../test";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));
jest.mock("../../support", () => ({
  assertAccountAdmin: jest.fn(() => () => undefined),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("createKeyPair", () => {
  let ctx: any;
  let keyPair: KeyPair;

  beforeEach(async () => {
    ctx = {
      logger,
      repository: await getTestRepository(),
    };
    keyPair = await ctx.repository.keyPair.create(getTestKeyPairEC());
  });

  afterEach(resetStore);

  test("should expire a key-pair using stringToDuration", async () => {
    await expect(
      expireKeyPair(ctx)({
        keyPairId: keyPair.id,
        expires: "1 days",
      }),
    ).resolves.toBe(undefined);

    expect(inMemoryStore).toMatchSnapshot();
  });

  test("should throw error if expires is wrong", async () => {
    await expect(
      expireKeyPair(ctx)({
        keyPairId: keyPair.id,
        expires: "wrong",
      }),
    ).rejects.toStrictEqual(expect.any(InvalidExpiryString));
  });
});
