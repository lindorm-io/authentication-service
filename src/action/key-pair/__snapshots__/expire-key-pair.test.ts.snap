// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createKeyPair should expire a key-pair using stringToDuration 1`] = `
Object {
  "authentication": Object {
    "key_pair": Object {
      "data": Array [
        Object {
          "algorithm": "ES512",
          "allowed": true,
          "created": 2020-01-01T07:00:00.000Z,
          "events": Array [
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "key_pair_created",
              "payload": Object {
                "algorithm": "ES512",
                "allowed": true,
                "expires": null,
                "passphrase": null,
                "privateKey": "-----BEGIN PRIVATE KEY-----
MIHuAgEAMBAGByqGSM49AgEGBSuBBAAjBIHWMIHTAgEBBEIBGma7xGZpaAngFXf3
mJF3IxZfDpI+6wU564K+eehxX104v6dZetjSfMx0rvsYX/s6cO2P3GE7R95VxWEk
+f4EX0qhgYkDgYYABAB8cBfDwCi41G4kVW4V3Y86nIMMCypYzfO8gYjpS091lxkM
goTRS3LM1p65KQfwBolrWIdVrbbOILASf06fQsHw5gEt4snVuMBO+LS6pesX9vA8
QT1LjX75Xq2InnLY1VToeNmxkuM+oDZgqHOYwzfUhu+zZaA5AuEkqPi47TA9iCSY
VQ==
-----END PRIVATE KEY-----
",
                "publicKey": "-----BEGIN PUBLIC KEY-----
MIGbMBAGByqGSM49AgEGBSuBBAAjA4GGAAQAfHAXw8AouNRuJFVuFd2POpyDDAsq
WM3zvIGI6UtPdZcZDIKE0UtyzNaeuSkH8AaJa1iHVa22ziCwEn9On0LB8OYBLeLJ
1bjATvi0uqXrF/bwPEE9S41++V6tiJ5y2NVU6HjZsZLjPqA2YKhzmMM31Ibvs2Wg
OQLhJKj4uO0wPYgkmFU=
-----END PUBLIC KEY-----
",
                "type": "ec",
              },
            },
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "key_pair_expires_changed",
              "payload": Object {
                "expires": 2020-01-02T07:00:00.000Z,
              },
            },
          ],
          "expires": 2020-01-02T07:00:00.000Z,
          "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
          "passphrase": null,
          "privateKey": "-----BEGIN PRIVATE KEY-----
MIHuAgEAMBAGByqGSM49AgEGBSuBBAAjBIHWMIHTAgEBBEIBGma7xGZpaAngFXf3
mJF3IxZfDpI+6wU564K+eehxX104v6dZetjSfMx0rvsYX/s6cO2P3GE7R95VxWEk
+f4EX0qhgYkDgYYABAB8cBfDwCi41G4kVW4V3Y86nIMMCypYzfO8gYjpS091lxkM
goTRS3LM1p65KQfwBolrWIdVrbbOILASf06fQsHw5gEt4snVuMBO+LS6pesX9vA8
QT1LjX75Xq2InnLY1VToeNmxkuM+oDZgqHOYwzfUhu+zZaA5AuEkqPi47TA9iCSY
VQ==
-----END PRIVATE KEY-----
",
          "publicKey": "-----BEGIN PUBLIC KEY-----
MIGbMBAGByqGSM49AgEGBSuBBAAjA4GGAAQAfHAXw8AouNRuJFVuFd2POpyDDAsq
WM3zvIGI6UtPdZcZDIKE0UtyzNaeuSkH8AaJa1iHVa22ziCwEn9On0LB8OYBLeLJ
1bjATvi0uqXrF/bwPEE9S41++V6tiJ5y2NVU6HjZsZLjPqA2YKhzmMM31Ibvs2Wg
OQLhJKj4uO0wPYgkmFU=
-----END PUBLIC KEY-----
",
          "type": "ec",
          "updated": 2020-01-01T07:00:00.000Z,
          "version": 1,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "id": 1,
          },
          "options": Object {
            "unique": true,
          },
        },
      ],
    },
  },
}
`;
